cardano-node/src/Cardano/Tracing/Tracers.hs

```diff
@@ -641,29 +645,76 @@ traceBlockFetchClientMetrics
+              fromIntegral cdf5s / fromIntegral size)
+
+
     bfTracer :: TraceLabelPeer remotePeer (TraceFetchClientState (Header blk)) -> IO ()
     bfTracer e@(TraceLabelPeer _ (CompletedBlockFetch p _ _ _ delay)) = do
       traceWith tracer e
-      let slotNo = case pointSlot p of
-                        Origin -> SlotNo 0 -- We don't care
-                        At s   -> s
-      fresh <- atomically $ do
-          lastSlotNo <- STM.readTVar lastSlotNoVar
-          if lastSlotNo < slotNo
-             then do
-                 STM.writeTVar lastSlotNoVar slotNo
-                 return True
-             else return False
-      when fresh $
-        sendEKGDirectDouble ekgDirect "cardano.node.metrics.blockfetchclient.blockdelay.s"
-            $ realToFrac delay
+      case pointSlot p of
+        Origin -> return () -- Nothing to do.
+        At slotNo -> do
+          (fresh, cdf1s, cdf2s, cdf5s) <- atomically $ do
+              slotMap <- STM.readTVar slotMapVar
+              case SMap.lookup slotNo slotMap of
+                   Nothing -> do
+                     let slotMap' = SMap.insert slotNo delay slotMap
+                     if SMap.size slotMap' > 1080 -- TODO k/2, should come from config file
+                        then
+                            let ((minSlotNo, minDelay), slotMap'') = SMap.deleteFindMin slotMap' in
+                            if minSlotNo == slotNo
+                               then return (False, 0, 0, 0) -- Nothing to do
+                               else do
+                                   _ <- updateCDFCounter minDelay (SMap.size slotMap'') (-1)
+                                   (cdf1s, cdf2s, cdf5s) <- updateCDFCounter delay
+                                       (SMap.size slotMap'') 1
+                                   STM.writeTVar slotMapVar slotMap''
+                                   return (True, cdf1s, cdf2s, cdf5s)
+                        else do
+                          (cdf1s, cdf2s, cdf5s) <- updateCDFCounter delay (SMap.size slotMap') 1
+                          STM.writeTVar slotMapVar slotMap'
+                          return (True, cdf1s, cdf2s, cdf5s)
+
+                   Just _ -> return (False, 0, 0, 0) -- dupe, we only track the first
+
+          when fresh $ do
+            sendEKGDirectDouble ekgDirect "cardano.node.metrics.blockfetchclient.blockdelay.s"
+                $ realToFrac delay
+            sendEKGDirectDouble ekgDirect "cardano.node.metrics.blockfetchclient.blockdelay.cdf1s"
+               cdf1s
+            sendEKGDirectDouble ekgDirect "cardano.node.metrics.blockfetchclient.blockdelay.cdf2s"
+               cdf2s
+            sendEKGDirectDouble ekgDirect "cardano.node.metrics.blockfetchclient.blockdelay.cdf5s"
+               cdf5s
+            when (delay > 5) $
+              sendEKGDirectCounter ekgDirect "cardano.node.metrics.blockfetchclient.lateblocks"
```

Comment:
It's painful that this is so slow.

Reply:
TODO added for when there is a faster way.

Reply:
@dcoutts, @karknu, would you expect the new APIs to provide you with a way to acquire the EKG object once, and then repeatedly post updates to it?

We only currently consider ways to improve the performance without paying the (synchronisation) cost of `modifyMVar` (or equivalents) -- but we're not exposing EKG objects directly.  We still expect to be performing a namespace map lookup per post.

Would you consider such a design unacceptable and would you like, instead, to (as mentioned above) operate on an once-acquired EKG object directly?